crypto/schnorr.c

```diff
@@ -0,0 +1,252 @@
+ * Schnorr and ECDSA.
+ */
+void init_rfc6979_schnorr(const uint8_t *priv_key, const uint8_t *hash,
+                          rfc6979_state *state) {
+  uint8_t hmac_data[SHA256_DIGEST_LENGTH + 16] = {0};
+
+  memcpy(hmac_data, hash, SHA256_DIGEST_LENGTH);
+  memcpy(hmac_data + SHA256_DIGEST_LENGTH, "Schnorr+SHA256  ", 16);
+
+  hmac_drbg_init(state, priv_key, 32, hmac_data, SHA256_DIGEST_LENGTH + 16);
+}
+
+int generate_k_schnorr(const ecdsa_curve *curve, bignum256 *k,
+                       rfc6979_state *state) {
+  int i = 0;
+
+  for (i = 0; i < 10000; i++) {
+    generate_k_rfc6979(k, state);
+    // If k is too big or too small, we don't like it
+    if (bn_is_zero(k) || !bn_is_less(k, &curve->order)) {
+      continue;
+    }
+
+    return 0;
+  }
+
+  return 1;
+}
+
+// e = H(Rx, pub_key, msg_hash)
+static void calc_e(const ecdsa_curve *curve, const bignum256 *Rx,
+                   const uint8_t pub_key[33], const uint8_t *msg_hash,
+                   bignum256 *e) {
+  uint8_t Rxbuf[32] = {0};
+  SHA256_CTX ctx = {0};
+  uint8_t digest[SHA256_DIGEST_LENGTH] = {0};
+
+  bn_write_be(Rx, Rxbuf);
+
+  sha256_Init(&ctx);
+  sha256_Update(&ctx, Rxbuf, sizeof(Rxbuf));
+  sha256_Update(&ctx, pub_key, 33);
+  sha256_Update(&ctx, msg_hash, SHA256_DIGEST_LENGTH);
+  sha256_Final(&ctx, digest);
+
+  bn_read_be(digest, e);
+  bn_fast_mod(e, &curve->order);
+  bn_mod(e, &curve->order);
+}
+
+int schnorr_sign(const ecdsa_curve *curve, const uint8_t *priv_key,
+                 const uint8_t *msg, const uint32_t msg_len, uint8_t *sign) {
+  uint8_t digest[SHA256_DIGEST_LENGTH] = {0};
+
+  double_sha256(msg, msg_len, digest);
+
+  return schnorr_sign_digest(curve, priv_key, digest, sign);
+}
+
+int schnorr_sign_digest(const ecdsa_curve *curve, const uint8_t *priv_key,
+                        const uint8_t *digest, uint8_t *sign) {
+  uint8_t pub_key[33] = {0};
+  curve_point R = {0};
+  bignum256 private_key_scalar = {0}, e = {0}, s = {0}, k = {0};
+  rfc6979_state rng = {0};
+
+  ecdsa_get_public_key33(curve, priv_key, pub_key);
+
+  // Compute k
+  init_rfc6979_schnorr(priv_key, digest, &rng);
+  if (generate_k_schnorr(curve, &k, &rng) != 0) {
+    memzero(&k, sizeof(k));
+    return 1;
+  }
+
+  // Compute R = k * G
+  point_multiply(curve, &k, &curve->G, &R);
+
+  // If R.y is not a quadratic residue, negate the nonce
+  bn_cnegate(is_non_quad_residue(&R.y, &curve->prime), &k, &curve->order);
+  bn_mod(&k, &curve->order);
+
+  bn_mod(&R.x, &curve->order);
+  bn_write_be(&R.x, sign);
+
+  // Compute e = H(Rx, pub_key, msg_hash)
+  calc_e(curve, &R.x, pub_key, digest, &e);
+
+  // Compute s = k + e * priv_key
+  bn_copy(&e, &s);
+  bn_read_be(priv_key, &private_key_scalar);
+  bn_multiply(&private_key_scalar, &s, &curve->order);
+  memzero(&private_key_scalar, sizeof(private_key_scalar));
+  bn_addmod(&s, &k, &curve->order);
+  memzero(&k, sizeof(k));
+  bn_fast_mod(&s, &curve->order);
```

Comment:
Calling `bn_fast_most` is not necessary since an output of `bn_addmod` is partly reduced.

Reply:
Removed in 1c7bd82